File: remove_all_adjacent_duplicates_in_string_II.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var removeDuplicates = function(s, k) {
    if (s.length < k) {
        return s;
    }

    const st = [];

    for (const c of s) {
        if (st.length && c === st[st.length - 1][0]) {
            st[st.length - 1][1]++;

            if (st[st.length - 1][1] === k) {
                st.pop();
            }
        } else {
            st.push([c, 1]);
        }
    }

    let res = "";
    for (const [char, count] of st) {
        res += char.repeat(count);
    }

    return res;    
};

File: remove_all_adjacent_duplicates_in_string_II.py
class Solution {
public:
    string removeDuplicates(string s, int k) {
        if (s.length() < k) {
            return s;
        }

        std::vector<std::pair<char, int>> st;

        for (char c : s) {
            if (!st.empty() && c == st.back().first) {
                st.back().second++;

                if (st.back().second == k) {
                    st.pop_back();
                }
            } else {
                st.emplace_back(c, 1);
            }
        }

        std::string res;
        for (const auto& [char_val, count] : st) {
            res.append(count, char_val);
        }

        return res;        
    }
};
